src/storage/txn/actions/prewrite.rs

```diff
@@ -75,6 +75,37 @@ pub fn prewrite<S: Snapshot>(
         return Ok((min_commit_ts, OldValue::Unspecified));
     }
 
+    // For keys that do not need pessimistic locks in a pessimistic async-commit transaction,
+    // we need to check the key has not been committed before if it is a retry request.
+    //
+    // It is to prevent the following case:
+    // The key was prewritten successfully before, but the response is lost. The client resends
+    // the same request after the transaction is resolved (to become committed). If we still
+    // write the lock, the client might commit these keys more than once.
+    //
+    // If the commit record is Rollback, it is still safe for us to write the Lock because the primary
+    // key of the transaction must also be rolled back, and this lock will be eventually rolled back.
+    // For simplicity, we don't handle this case specially, making it the same behavior as the Rollback
+    // record is collapsed.
+    if txn_props.is_pessimistic()
+        && !is_pessimistic_lock
+        && txn_props.retry_request
+        && matches!(txn_props.commit_kind, CommitKind::Async(..))
```

Comment:
Does 1PC have any related problem?

Reply:
emm....  seems not

Reply:
Emm, I find something wrong. When encountering `PessimisticLockNotFound`, the buffered `modifies` should be cleared. Otherwise, the problem can happen on 1PC.

Reply:
Ah, it won't make any problem because the final commit ts is correct, the only problem is that the same record is rewritten once...
Though it does not affect correctness, I still clear the `modifies` and add one more test for it. Please take a look.

Reply:
Emmm.. The change looks fine to me, but I think I didn't fully understand it. In which case is it possible that a replayed 1PC request buffered some modifies?

Reply:
For example, the first prewritten key does not need a pessimistic lock, so it saves its 1PC memory lock. Then the second key is a row key which needs a pessimistic lock, it then encounters PessimisticLockNotFound error. But later, we find the transaction is actually committed so no error is returned. We continue the 1PC process and convert the previous 1PC lock to a Write mutation.

Reply:
Do we need to call the `txn.clear()` is not called here?

Reply:
> Do we need to call the txn.clear() here?
Not necessary, but good to have. Now I added it.